core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -99,20 +140,54 @@ public void registerCacheEventListener(CacheEventListener<? super K, ? super V>
    */
   @Override
   public void deregisterCacheEventListener(CacheEventListener<? super K, ? super V> listener) {
```

Comment:
Objective – deregister listener, decrement ordered count (if necessary). Remove listener from respective sync/async set and disable threadlocal if no listener left

What it internally does:
•	Create a listener wrapper that has all necessary information of the listener.
•	Check if listener present in either of the sync/async set and remove, also reduce orderedListener count 
•	Finally disable EventThreadLocal to a noop if no listeners are left.

Why it should work:
A `removed` boolean is maintained that helps to track if the listener did get removed from the concerned listener set. Along with the removal from set, if the listener was ordered, a decrement in the ordered listener count is also done. Also if no listeners are left, the threadLocal is switched to a no-op under lock.